src/storage.rs

```diff
@@ -0,0 +1,291 @@
+use crate::{
+    poly::{Poly, PolyMatrix},
+    utils::{block_size, debug_mem, log_mem},
+};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use tokio::{
+    runtime::Handle,
+    sync::{mpsc, oneshot},
+};
+
+#[derive(Debug)]
+pub struct SerializedBlock {
+    filename: String,
+    data: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct SerializedMatrix {
+    pub id: String,
+    pub blocks: Vec<SerializedBlock>,
+}
+
+/// Storage task sent to the I/O worker thread
+struct StorageTask {
+    serialized_data: Vec<u8>,
+    path: PathBuf,
+    id: String,
+    completion_sender: oneshot::Sender<Result<(), std::io::Error>>,
+}
+
+/// Global storage service for handling all matrix I/O operations
+pub struct StorageService {
+    task_sender: mpsc::UnboundedSender<StorageTask>,
+    _worker_handle: tokio::task::JoinHandle<()>,
+}
+
+impl StorageService {
+    /// Creates a new storage service with configurable concurrent writes
+    pub fn new(max_concurrent_writes: usize) -> Self {
+        let (task_sender, mut task_receiver) = mpsc::unbounded_channel::<StorageTask>();
+
+        let _worker_handle = tokio::spawn(async move {
+            let semaphore = Arc::new(tokio::sync::Semaphore::new(max_concurrent_writes));
+
+            while let Some(task) = task_receiver.recv().await {
+                let semaphore = semaphore.clone();
+
+                // Spawn individual I/O tasks up to the concurrency limit
+                tokio::spawn(async move {
+                    let _permit =
+                        semaphore.acquire().await.expect("Semaphore should not be closed");
```

Comment:
What's the difference between using `semaphore` and directly setting the bound to the channel in L44?

Reply:
Also do you plan to modify the implementation later to adjust the bound dynamically based on the RAM usage?